Automattic/newspack-plugin
# feat(ras): sync class

### All Submissions:






### Changes proposed in this Pull Request:

Following #3359 and #3360, this PR continues to work on centralizing how we perform reader sync. Sync data is currently handled in a few different places:

 - [`Newspack\WooCommerce_Connection`](https://github.com/Automattic/newspack-plugin/blob/b5f0f37395485add7e643c6daec07cbad1bb5e4b/includes/reader-revenue/woocommerce/class-woocommerce-connection.php): generating contact data from customer or order ID
 - [`Newspack\Newspack_Newsletters`](https://github.com/Automattic/newspack-plugin/blob/b5f0f37395485add7e643c6daec07cbad1bb5e4b/includes/plugins/class-newspack-newsletters.php): managing the metadata and normalizing contact data in syncs

The `Newspack\Newspack_Newsletters` class has been deprecated and its metadata handling moved to [`Newspack\Reader_Activation\Sync\Metadata`](https://github.com/Automattic/newspack-plugin/blob/5b06af1e401594b1cd34c53949778495eaec0c21/includes/reader-activation/sync/class-metadata.php). A few notable changes in its behavior:
 - The `normalize_contact_data` is no longer applied as a filter to `newspack_newsletters_contact_data`. It's now called inside `Newspack\Reader_Activation\ESP_Sync::sync()`, right after the `newspack_esp_sync_contact` filter and before the `Newspack_Newsletters_Contacts::upsert()` call:

https://github.com/Automattic/newspack-plugin/blob/5b06af1e401594b1cd34c53949778495eaec0c21/includes/reader-activation/sync/class-esp-sync.php#L97-L109

 - The `normalize_contact_data` should no longer handle provider-specific adjustments, like Mailchimp's First Name and Last Name merge fields. See https://github.com/Automattic/newspack-newsletters/pull/1628
 - [Log when a metadata key is being ignored for not being allowed in](https://github.com/Automattic/newspack-plugin/blob/5b06af1e401594b1cd34c53949778495eaec0c21/includes/reader-activation/sync/class-metadata.php#L381)
 - Allow UTM keys to be sent raw, like `signup_page_utm_campaign`, and be formatted during normalization

From `Newspack\WooCommerce_Connection`, the methods `should_sync_order()`, `get_order_metadata()`, `get_contact_from_customer()`, and `get_contact_from_order()` moved to a new [`Newspack\Reader_Activation\Sync\WooCommerce`](https://github.com/Automattic/newspack-plugin/blob/5b06af1e401594b1cd34c53949778495eaec0c21/includes/reader-activation/sync/class-woocommerce.php) class.

The metadata generation on `get_order_metadata()` and `get_contact_from_customer()` no longer formats the keys with the prefixes from `Metadata`. Instead, it uses the raw keys to be later transformed via `normalize_contact_data`. This is also the case in other places where metadata is populated, like the [`Newspack\Data_Events\Connectors\ESP_Connector`](https://github.com/Automattic/newspack-plugin/blob/5b06af1e401594b1cd34c53949778495eaec0c21/includes/data-events/connectors/class-esp-connector.php) and [the Teams for Memberships integration](https://github.com/Automattic/newspack-plugin/blob/32584d801e5ae74dcc22c88a234ab22f755f13e1/includes/plugins/class-teams-for-memberships.php#L99).

### How to test the changes in this Pull Request:

All sync methods should be tested. First, repeat the testing steps from #3359 and #3360.

Finally, we should also test the UTM integration:

1. In a fresh session, visit your site with UTM parameters: `?utm_source=website&utm_medium=homepage&utm_campaign=header`
2. Donate and confirm the sync goes through including the UTM parameters

### Other information:
